### ISSUE TITLE
Long Term Data on RPI2 failing with certain selections

### ISSUE THREAD

**Issue by pchristod** [pi-hole/pi-hole#1845]
**In raising this issue, I confirm the following:** 

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

`1`

---
**Expected behaviour:**

`The Data I selected`

**Actual behaviour:**

`The more Data selected the longer it takes for the Web Interface to compute the Data, selecting a Full Month from the pre defined "This Month" "Last Month" will crash FTL in the Background somehow (or the Database), the Loading Circle will just spin indefinetly, FTL itself will still work but you're not able to compute any more Long Term Data (even just for today) until FTL is stopped and restarted. The Error shown then is this: https://imgur.com/SBaBDtj`

**Steps to reproduce:**

`At least on a RPI 2 selecting a huge timefrime will make it crash, until "Last 30 Days" or even "last Month" (which is basically the same timeframe wise" it works sometimes, "This Year" or even "All Time" will make it crash for sure. I believe it is something with the Hardware or Memory which runs out?`

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

`xnszmambi1`

**Troubleshooting undertaken, and/or other relevant information:**

`Deleted Browser Cache, Repaired Pihole and tried another Browser/Machine`

**Comment by Mcat12** [comment]
Have you tried refreshing the page after it is unable to show the date range?

**Comment by pchristod** [comment]
Yes, the page is refreshed normally with the exception that you can't compute any data even for today /1 Day. I need to stop and restart FTL for it to work again.

**Comment by Mcat12** [comment]
I can't duplicate being unable to make further requests. That is, once one fails I can try another and it works without having to restart FTL.

**Comment by pchristod** [comment]
Are you using a Pic too? May be also something with the Database size!?
I have tried it a dozen times, after a failed request I can't make another one, probably because the first didn't ever finish (indicated by the ever loading circle), so it wont start another request.
Also have you tried to show a lot of data (time wise)?

**Comment by Mcat12** [comment]
I am actually using a Tinkerboard, but that is equivalent to a Pi 3. Even after trying to get "All Time" and failing, I could choose "Yesterday" and it would work. I didn't see anything in your logs to indicate that something crashed, or that this wouldn't work on your side. FTL did not crash and didn't show any errors.

**Comment by pchristod** [comment]
Sorry, Autocorrect, I meant Pi2!
I uploaded another Debug right after a failed attempt: r9jbuw672h

Also I made a small Video on my Phone (simplest way to Screenrecord) but it actually behaves exactly the same on a PC or any other Device. Sorry for the Tint, it's the Night Mode. Actually what I've done:
Selected "Today" to show it's working, selected "All Time" and waited a few seconds ( this is just to make the Video shorter, actually I've waited around 10 Minutes one time to see if something was going to be displayed eventually), after that I selected "Today" again to see nothing is coming up. Refreshed every time, you can see it on the Circle on Top. Switching to SSH and restarting/stopping FTL, refresh and selecting "Today" again makes it work.
Here's the Link, it 1:30 Minutes: https://www.youtube.com/watch?v=7AgKqcz-KIY&feature=youtu.be

**Comment by Mcat12** [comment]
PHP holds a read lock on the database when you execute the long select command, and subsequent requests won't go through until it finishes reading the database. By restarting FTL, I have a feeling FTL does something to the database to remove any locks on it.

There's a 30 second timeout, and my machine was sending an error quickly on large queries because it couldn't allocate enough memory. On some smaller queries though, it did take about 20 seconds. We're working on improving performance, but for now it is expected that it will take a long time for large queries.

**Comment by pchristod** [comment]
I see. But what I don't get is why there is never a timeout for me. You say after 30 seconds a query should be finished or throw an error but I could let it load for Minutes without any result. That a lot of data takes memory and time is only logical.
Also thank you for your continued support

**Comment by Mcat12** [comment]
Thanks. The default timeout is usually set to 30 seconds, but I remember now that there is a settings override we have for the web interface located at `/var/www/html/admin/.user.php.ini`. If you edit that file and modify `max_execution_time` to 30 seconds, does it time out quicker? Also, you know when the request fails by when the error alert comes up. After that alert comes up, it will still say processing. To be sure, you can check the Network tab in the dev tools of your browser to see when the request finishes.

**Comment by s1awa** [comment]
Have RPI2. Same issue. If querying too much long term data I get `500 Internal Server Error` on the API call

/admin/db_queries.php

Last 30 days -> 500 Internal Server Error
the server recovers and still can query 7 days

All time -> 500 Internal Server Error
any API call fails now with `500 Internal Server Error`, `sudo service pihole-FTL restart` required to fix

**Comment by DL6ER** [comment]
I should have pointed out here that I found the bug in `FTL` and fixed it here: https://github.com/pi-hole/FTL/pull/185

TL;DR: When the database is locked due to `PHP` still trying to get the requested data, `FTL` wasn't able to close the database handle as it already had prepared to write new data into the table. Because we didn't check for the case that closing the database could fail, this got unnoticed. It will be fixed in the next release of `FTL`.

**Comment by pchristod** [comment]
Awesome, thank you!
Sorry for not reporting back, Holidays with Family etc. you know how it is, haven't had the Chance to try what @Mcat12 suggested but it seems obsolete now anyways :)

**Comment by s1awa** [comment]
As you can see in my screenshot API queries timeout after 30 seconds. (PHP default maximum execution time limit is 30 seconds)

Would it work if we increase [set_time_limit](https://secure.php.net/manual/en/function.set-time-limit.php) for big queries?

**Comment by Mcat12** [comment]
We'd rather work towards increasing the performance of the API than increasing the time limit (we have already increased it via [this](https://github.com/pi-hole/AdminLTE/blob/master/.user.php.ini) file in the past).

**Comment by s1awa** [comment]
Just tested new FTL v2.13.2
'Times out' with status 500 after 30 seconds but doesn't lock up anymore.

**Comment by DL6ER** [comment]
That's expected (the "PHP is too slow to process your request" issue cannot be solved from within FTL).

You should increase `max_execution_time` if you want to use such requests. We will increase it to 10min (only for the DB part of the API) in the next release.
